### PR TITLE
feat: notification creation

### PR DESCRIPTION
**Pull request by FerranAD**
This closes: #20.

### CODE REVIEW

File: src/main/kotlin/notification/application/NotificationCreator.kt [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+import note.domain.exceptions.NonExistentNoteException
+import notification.domain.Notification
+import notification.domain.NotificationRepository
+import notification.domain.exceptions.AlreadyConfiguredNotification
+import shared.domain.Identifier
+import shared.domain.exceptions.InvalidUUIDException
+
+class NotificationCreator(
+    private val repository: NotificationRepository,
+    private val noteSearcher: NoteRepository
```
Comment (Pablito2020) [inline src/main/kotlin/notification/application/NotificationCreator.kt:15]: Should be named noteRepository, since noteSearcher is the name of the use case:
```suggestion
    private val noteRepository: NoteRepository
```

File: src/main/kotlin/notification/domain/Notification.kt [hunk 2]
```diff
@@ -1,5 +1,6 @@
 package notification.domain
 
+import deadline.domain.Time
```
Comment (Pablito2020) [inline src/main/kotlin/notification/domain/Notification.kt:3]: I will create an issue for extracting the Time class to shared, since it is a class that can be needed by more than one module.

File: src/test/kotlin/notification/application/NotificationCreatorTest.kt [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+        assertThrows<NonExistentNoteException> { notificationCreator.create(notificationPrimitives) }
+    }
+
+    @Test
+    fun `If a notification is already configured for the given time, throw an exception`() {
+        val notification = NotificationMother.getValidNotification()
+        val noteId = NotificationMother.getNoteIdentifierFrom(notification)
+        val notificationPrimitives = NotificationMother.getPrimitivesFrom(notification)
+
```
Comment (Pablito2020) [inline src/test/kotlin/notification/application/NotificationCreatorTest.kt:44]: Here, it makes sense to first create a valid note, and then assign a valid notification for it. (Since a note can live without a notification but a notification can't live without a note)
```kotlin
       val note = NoteMother.getValidNote()
       val notification = NotificationMother.getValidNotificationFor(note)
```
So the getNoteIdentifierFrom(notification) isn't needed, and NoteMother.getValidNoteFromIdentifier isn't needed too

File: src/test/kotlin/notification/application/NotificationCreatorTest.kt [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    @Test
+    fun `If a notification is not configured for the given time, save it to the repository`() {
+        val notification = NotificationMother.getValidNotification()
+        val noteId = NotificationMother.getNoteIdentifierFrom(notification)
+        val notificationPrimitives = NotificationMother.getPrimitivesFrom(notification)
+
+        Mockito.`when`(noteSearcher.search(noteId)).thenReturn(NoteMother.getValidNoteFromIdentifier(noteId))
+        Mockito.`when`(notificationRepository.search(noteId)).thenReturn(emptyList<Notification>())
+        val result = notificationCreator.create(notificationPrimitives)
+
```
Comment (Pablito2020) [inline src/test/kotlin/notification/application/NotificationCreatorTest.kt:60]: Same as the other case, it is better to create a new valid note  and then assign a notification (that can be a good one or a bad one) to it.

File: src/test/kotlin/note/mothers/NoteMother.kt [hunk 5]
```diff
@@ -28,4 +29,10 @@ object NoteMother {
     }
 
     fun getIdentifierFrom(note: Note) = note.id
+
+    fun getValidNoteFromIdentifier(identifier: Identifier) = Note(
```
Comment (Pablito2020) [inline src/test/kotlin/note/mothers/NoteMother.kt:33]: Isn't really needed, check the comment I told you inside the NotificationCreatorTest file.

**Review by Pablito2020 — CHANGES_REQUESTED** [review]
Some renaming and test objects instantiation.

**Review by oriolagobat — APPROVED** [review]

File: src/test/kotlin/note/mothers/NoteMother.kt [hunk 6]
```diff
@@ -28,4 +28,6 @@ object NoteMother {
     }
 
     fun getIdentifierFrom(note: Note) = note.id
+
+    fun getValidNote() = getValidNoteWithDescription()
```
Comment (Pablito2020) [inline src/test/kotlin/note/mothers/NoteMother.kt:32]: Why is this needed? A function that calls another function without any new parameter isn't a good sign.
In the test cases you can create a valid note with description and it will do the job

**Review by Pablito2020 — CHANGES_REQUESTED** [review]

**Review by Pablito2020 — APPROVED** [review]
